appsrc/dcfile/rawtodc.man
.TH RAWTODC 1 "9 December 2005" "DICOM PS3" "Create DICOM from raw file"
.SH NAME
rawtodc \- ACR/NEMA DICOM PS3 ... Create DICOM from raw file
.SH SYNOPSIS
.HP 10
.B rawtodc
.B \-rows|\-height|\-h " rows"
.B \-columns|\-width|\-w " cols"
.B \-bits|\-depth|\-d " bits"
[
.B \-frames " frames"
]
[
.B \-little|\-l|\-big|\-b
]
[
.B \-samples " samples"
]
[
.B \-gray|\-grey|\-rgb|\-argb|\-hsv|\-cmyk
]
[
.B \-photometric-interpretation " xxx"
]
[
.B \-color-by-pixel|\-color-by-plane
]
[
.B \-imagepixelmoduleonly
]
[
.B \-v|verbose
]
.so man1/genin.so
.so man1/optout.so
.SH DESCRIPTION
.LP
.B rawtodc
reads the named raw input file and converts it to a DICOM Secondary Capture Image Storage SOP Class instance stored in a DICOM file. Only uncompressed raw images with one sample per byte or word are supported, though multiple samples per pixel are allowed. The planar configuration can be specified but not re-ordered.
.LP
During the translation, specific attributes may be deleted, added or
replaced, class and instance unique identifiers generated or removed,
group lengths added, or private attributes removed, as specified by
the replacement options described in dcintro(1). This allows the
specification of additional DICOM attributes, such as the patient's
name, necessary to make a complete and useful SC instance.
.SH OPTIONS
The verbose output goes to standard error.
.PP
.PP
The basic switches are described in dcintro(1). Options specific to this program are:
.TP
.BI \-rows|\-height|\-h " rows"
.RS
Required.
.RE
.TP
.BI \-columns|\-width|\-w " cols"
.RS
Required.
.RE
.TP
.BI \-bits|\-depth|\-d " bits"
.RS
Required.
.RE
.TP
.BI \-little|\-big
.RS
Required if bit depth is greater than 8.
.RE
.TP
.BI \-frames " frames"
.RS
Defaults to 1.
.RE
.TP
.BI \-samples " samples"
.RS
Defaults to appropriate value for photometric interpretation, or 1 if none specified. Only needs to be specified for unusual photometric interpretations.
.RE
.TP
.BI \-gray|\-grey|\-rgb|\-argb|\-hsv|\-cmyk
.RS
Recognized photometric interpretations, defaults to gray if no samples specified or samples is 1, and rgb if samples is 3.
.RE
.TP
.BI \-photometric-interpretation " xxx"
.RS
Explicitly specify string value for photometric interpretation attribute; will need value for samples.
.RE
.TP
.BI \-color-by-pixel|\-color-by-plane
.RS
Specifies the planar configuration. Required if samples not 1 (explicitly or deduced from photometric interpretation). No default.
.RE
.TP
.BI \-imagepixelmoduleonly
.RS
Write only the image pixel data module and not the basic mandatory secondary capture and composite IOD modules as well
(e.g., prior to dcmerge'ing with an existing SOP instance that has its pixel data deleted); use the -nodisclaimer switch
to avoid generating Image Comments and the -n switch to avoid adding a meta-information header (will fail if you do not
since there is no SOP Class UID to use in the meta-information header).
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
# Create a gray image from 16 bit big-endian ...
.RE
% rawtodc -rows 256 -columns 256 -bits 16 -big \\
.RE
          /tmp/test.big /tmp/test.dc3
.RE

.RE
# Create a gray image from 8 bit ...
.RE
% rawtodc -rows 256 -columns 256 -bits 8 \\
.RE
          /tmp/test.byte /tmp/test.dc3
.RE
\ 
.RE
# Create an rgb color-by-pixel image from 8 bit ...
.RE
% rawtodc -v -rows 256 -columns 256 -bits 8 \\
.RE
             -rgb -color-by-pixel \\
.RE
          /tmp/test.byte3 /tmp/test.dc3
.RE
 ...
.RE
(0x0028,0x0002) US Samples per Pixel [0x0003] 
.RE
(0x0028,0x0004) CS Photometric Interpretation <RGB > 
.RE
(0x0028,0x0006) US Planar Configuration [0x0000] 
.RE
 ...
.RE
\ 
.RE
# Create a gray image from 12 bit little endian ...
.RE
% rawtodc -v -rows 256 -columns 256 -bits 12 -little \\
.RE
          /tmp/test.little /tmp/test.dc3
.RE
Converting ...
.RE
******** As read ... ********
.RE
(0x0008,0x0020) DA Study Date <> 
.RE
(0x0008,0x0030) TM Study Time <> 
.RE
(0x0008,0x0050) SH Accession Number <> 
.RE
(0x0008,0x0060) CS Modality <OT> 
.RE
(0x0008,0x0064) CS Conversion Type <WSD > 
.RE
(0x0008,0x0070) LO Manufacturer <> 
.RE
(0x0008,0x0090) PN Referring Physician's Name <^^^^> 
.RE
(0x0010,0x0010) PN Patient's Name <^^^^>
.RE
(0x0010,0x0020) LO Patient's ID <> 
.RE
(0x0010,0x0030) DA Patient's Birth Date <> 
.RE
(0x0010,0x0040) CS Patient's Sex <> 
.RE
(0x0020,0x0010) SH Study ID <> 
.RE
(0x0020,0x0011) IS Series Number <> 
.RE
(0x0020,0x0013) IS Image Number <> 
.RE
(0x0020,0x0020) CS Patient Orientation <> 
.RE
(0x0028,0x0002) US Samples per Pixel [0x0001] 
.RE
(0x0028,0x0004) CS Photometric Interpretation <MONOCHROME2> 
.RE
(0x0028,0x0008) IS Number of Frames <1 > 
.RE
(0x0028,0x0010) US Rows [0x0100] 
.RE
(0x0028,0x0011) US Columns [0x0100] 
.RE
(0x0028,0x0100) US Bits Allocated [0x0010] 
.RE
(0x0028,0x0101) US Bits Stored [0x000c] 
.RE
(0x0028,0x0102) US High Bit [0x000b] 
.RE
(0x0028,0x0103) US Pixel Representation [0x0000] 
.RE
(0x7fe0,0x0010) OX Pixel Data VR=<OW>   VL=<0x20000>  
.RE
******** As written ... ********
.RE
(0x0002,0x0000) UL Meta Element Group Length [0x000000a4] 
.RE
(0x0002,0x0001) OB File Meta Information Version \\
.RE
                   [0x00,0x01] 
.RE
(0x0002,0x0002) UI Media Storage SOP Class UID \\
.RE
                   <1.2.840.10008.5.1.4.1.1.7> 
.RE
(0x0002,0x0003) UI Media Storage SOP Instance UID \\
.RE
                   <0.0.0.0.1.0.0.0.830260380> 
.RE
(0x0002,0x0010) UI Transfer Syntax UID \\
.RE
                   <1.2.840.10008.1.2> 
.RE
(0x0002,0x0012) UI Implementation Class UID  <0.0.0.0> 
.RE
(0x0002,0x0013) SH Implementation Version Name \\
.RE
                   <NOTSPECIFIED> 
.RE
(0x0002,0x0016) AE Source Application Entity Title \\
.RE
                   <NOTSPECIFIED> 
.RE
(0x0008,0x0016) UI SOP Class UID \\
.RE
                   <1.2.840.10008.5.1.4.1.1.7> 
.RE
(0x0008,0x0018) UI SOP Instance UID \\
.RE
                   <0.0.0.0.1.0.0.0.830260380> 
.RE
(0x0008,0x0020) DA Study Date <> 
.RE
(0x0008,0x0030) TM Study Time <> 
.RE
(0x0008,0x0050) SH Accession Number <> 
.RE
(0x0008,0x0060) CS Modality <OT> 
.RE
(0x0008,0x0064) CS Conversion Type <WSD > 
.RE
(0x0008,0x0070) LO Manufacturer <> 
.RE
(0x0008,0x0090) PN Referring Physician's Name <^^^^> 
.RE
(0x0010,0x0010) PN Patient's Name 			<^^^^> 
.RE
(0x0010,0x0020) LO Patient's ID <> 
.RE
(0x0010,0x0030) DA Patient's Birth Date <> 
.RE
(0x0010,0x0040) CS Patient's Sex <> 
.RE
(0x0020,0x000d) UI Study Instance UID \\
.RE
                   <0.0.0.0.2.0.830260380> 
.RE
(0x0020,0x000e) UI Series Instance UID \\
.RE
                   <0.0.0.0.3.0.0.830260380> 
.RE
(0x0020,0x0010) SH Study ID <> 
.RE
(0x0020,0x0011) IS Series Number <> 
.RE
(0x0020,0x0013) IS Image Number <> 
.RE
(0x0020,0x0020) CS Patient Orientation <> 
.RE
(0x0028,0x0002) US Samples per Pixel [0x0001] 
.RE
(0x0028,0x0004) CS Photometric Interpretation <MONOCHROME2> 
.RE
(0x0028,0x0008) IS Number of Frames <1 > 
.RE
(0x0028,0x0010) US Rows [0x0100] 
.RE
(0x0028,0x0011) US Columns [0x0100] 
.RE
(0x0028,0x0100) US Bits Allocated [0x0010] 
.RE
(0x0028,0x0101) US Bits Stored [0x000c] 
.RE
(0x0028,0x0102) US High Bit [0x000b] 
.RE
(0x0028,0x0103) US Pixel Representation [0x0000] 
.RE
(0x7fe0,0x0010) OX Pixel Data VR=<OW>   VL=<0x20000>  
.RE
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dccp(1) ,
.BR dctoraw(1) ,
.BR pnmtodc(1) ,
.BR dciodvfy(1) ,
.BR dcintro(1) ,
.BR dcmerge(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS


CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/rawtodc.cc
#include "attrtype.h"
#include "attrothr.h"
#include "attrmxls.h"
#include "transynu.h"
#include "elmconst.h"
#include "mesgtext.h"
#include "ioopt.h"
#include "dcopt.h"
#include "rawsrc.h"

int
main(int argc,char **argv)
{
	bool bad=false;

	GetNamedOptions 	options(argc,argv);
	InputOptions		input_options(options);
	DicomOutputOptions 	dicom_output_options(options);

	bool verbose=options.get("verbose") || options.get("v");

	bool imagepixelmoduleonly=options.get("imagepixelmoduleonly");

	long offset=0;
	(void)options.get("offset",offset);

	unsigned rows=0;
	if (!options.get("rows",rows) && !options.get("height",rows) && !options.get("h",rows)) {
		cerr << EMsgDC(NeedOption) << " - rows" << endl;
		bad=true;
	}
	
	unsigned cols=0;
	if (!options.get("columns",cols) && !options.get("width",cols) && !options.get("w",cols)) {
		cerr << EMsgDC(NeedOption) << " - columns" << endl;
		bad=true;
	}
	
	unsigned frames=1;
	options.get("frames",frames);
	
	unsigned bits=0;
	if (!options.get("bits",bits) && !options.get("depth",bits) && !options.get("d",bits)) {
		cerr << EMsgDC(NeedOption) << " - bits" << endl;
		bad=true;
	}

	bool littleendian=options.get("little") || options.get("l");
	bool bigendian=options.get("big") || options.get("b");

	if (littleendian && bigendian) {
		cerr << EMsgDC(OptionsIncompatible) << " - big & little" << endl;
		bad=true;
	}
	else if (bits > 8 && !(littleendian || bigendian)) {
		cerr << EMsgDC(OptionsIncompatible) << " - bits > 8 && !(little|big)" << endl;
		bad=true;
	}
	else if (bits <= 8 && (littleendian || bigendian)) {
		cerr << EMsgDC(OptionsIncompatible) << " - bits <= 8 & (little|big)" << endl;
		bad=true;
	}

	if (bits < 1 || bits > 16) {
		cerr << EMsgDC(OptionUnsupported) << " - bits = " << dec << bits << endl;
		bad=true;
	}

	unsigned samples;
	bool samplespresent=options.get("samples",samples);

	const char *photometricInterpretation=0;

	int graypresent=(options.get("gray") || options.get("grey")) ? 1 : 0;
	int rgbpresent=options.get("rgb") ? 1 : 0;
	int argbpresent=options.get("argb") ? 1 : 0;
	int hsvpresent=options.get("hsv") ? 1 : 0;
	int cmykpresent=options.get("cmyk") ? 1 : 0;
	int photometricpresent=options.get("photometric-interpretation",photometricInterpretation) ? 1 : 0;

	int sumofphotometric=graypresent+rgbpresent+argbpresent+hsvpresent+cmykpresent+photometricpresent;

	if (sumofphotometric == 0) {
		if (samplespresent) {
			if (samples == 1) {
				photometricInterpretation="MONOCHROME2";
			}
			else if (samples == 3) {
				photometricInterpretation="RGB";
			}
		}
		else {
			photometricInterpretation="MONOCHROME2";
			samples=1;
		}
	}
	else if (sumofphotometric > 1) {
		cerr << EMsgDC(OptionsIncompatible) << " - gray,rgb,argb,hsv,cmyk,photometric-interpretation" << endl;
		bad=true;
	}
	else {
		if (graypresent) {
			if (samplespresent && samples != 1) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 1 & gray" << endl;
				bad=true;
			}
			else
				samples=1;
			photometricInterpretation="MONOCHROME2";
		}
		else if (rgbpresent) {
			if (samplespresent && samples != 3) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 3 & rgb" << endl;
				bad=true;
			}
			else
				samples=3;
			photometricInterpretation="RGB";
		}
		else if (argbpresent) {
			if (samplespresent && samples != 4) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 4 & argb" << endl;
				bad=true;
			}
			else
				samples=4;
			photometricInterpretation="ARGB";
		}
		else if (hsvpresent) {
			if (samplespresent && samples != 3) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 3 & hsv" << endl;
				bad=true;
			}
			else
				samples=3;
			photometricInterpretation="HSV";
		}
		else if (cmykpresent) {
			if (samplespresent && samples != 4) {
				cerr << EMsgDC(OptionsIncompatible) << " - samples != 4 & cmyk" << endl;
				bad=true;
			}
			else
				samples=4;
			photometricInterpretation="CMYK";
		}
		else if (photometricpresent) {
			if (!samplespresent) {
				cerr << EMsgDC(OptionRequired) << " - samples" << endl;
				bad=true;
			}
		}
	}

	Assert(photometricInterpretation);
	Assert(samples);

	bool bypixelpresent=options.get("color-by-pixel");
	bool byplanepresent=options.get("color-by-plane");

	unsigned planarConfiguration;

	if (bypixelpresent && byplanepresent) {
		cerr << EMsgDC(OptionsIncompatible) << " - color-by-pixel & color-by-plane" << endl;
		bad=true;
	}
	if (samples == 1) {
		if (bypixelpresent || byplanepresent) {
			cerr << EMsgDC(OptionsIncompatible) << " - (color-by-pixel|color-by-plane) & samples = 1" << endl;
			bad=true;
		}
	}
	else {
		if (bypixelpresent)
			planarConfiguration=0;
		else if (byplanepresent)
			planarConfiguration=1;
		else {
			cerr << EMsgDC(OptionRequired) << " - color-by-pixel | color-by-plane" << endl;
			bad=true;
		}
	}
			
	input_options.done();
	dicom_output_options.done();
	options.done();

	InputOpenerFromOptions input_opener(
		options,input_options.filename,cin);
	DicomOutputOpenerFromOptions output_opener(
		options,dicom_output_options.filename,cout);

	cerr << input_options.errors();
	cerr << dicom_output_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();
	cerr << output_opener.errors();

	if (!input_options.good()
	 || !dicom_output_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !output_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< input_options.usage()
			<< dicom_output_options.usage()
			<< " -rows|-height|-h rows"
			<< " -columns|-width|-w columns"
			<< " -bits|-depth|-d bits"
			<< " [-frames frames]"
			<< " [-little|-l|-big|-b]"
			<< " [-samples n]"
			<< " [-offset bytes]"
			<< " [[-gray|-grey|-rgb|-argb|-hsv|-cmyk]|[-photometric-interpretation xxx]]"
			<< " [-color-by-pixel|-color-by-plane]"
			<< " [-imagepixelmoduleonly]"
			<< " [-v|-verbose]"
			<< " [" << MMsgDC(InputFile)
				<< "[" << MMsgDC(OutputFile) << "]]"
			<< " <" << MMsgDC(InputFile)
			<< " >" << MMsgDC(OutputFile)
			<< endl;
		exit(1);
	}

	istream in(input_opener);
	DicomOutputStream dout(*(ostream *)output_opener,
		dicom_output_options.transfersyntaxuid,
		dicom_output_options.usemetaheader,
		dicom_output_options.useimplicitmetaheader,
		dicom_output_options.addtiff);

	ManagedAttributeList list;

	TransferSyntax transfersyntax(
		dicom_output_options.transfersyntaxuid
		? dicom_output_options.transfersyntaxuid
		: DefaultTransferSyntaxUID);

	TextOutputStream log (cerr);
	if (verbose) log << "Converting ..." << endl;

	Uint16 bitsAllocated=((bits-1)/8+1)*8;
	Uint16 bitsStored=bits;
	Uint16 highBit=bits-1;

	list+=new UnsignedShortAttribute(TagFromName(BitsAllocated),bitsAllocated);
	list+=new UnsignedShortAttribute(TagFromName(BitsStored),bitsStored);
	list+=new UnsignedShortAttribute(TagFromName(HighBit),highBit);
	list+=new UnsignedShortAttribute(TagFromName(Rows),rows);
	list+=new UnsignedShortAttribute(TagFromName(Columns),cols);
	list+=new IntegerStringAttribute(TagFromName(NumberOfFrames),(Uint16)frames);
	list+=new UnsignedShortAttribute(TagFromName(SamplesPerPixel),samples);
	list+=new UnsignedShortAttribute(TagFromName(PixelRepresentation),0u);
	list+=new CodeStringAttribute(TagFromName(PhotometricInterpretation),photometricInterpretation);
	if (samples > 1) list+=new UnsignedShortAttribute(TagFromName(PlanarConfiguration),planarConfiguration);

	if (!imagepixelmoduleonly) {
		// various Type 1 and Type 2 attributes for mandatory SC modules ...

		list+=new PersonNameAttribute(TagFromName(PatientName),"^^^^");
		list+=new LongStringAttribute(TagFromName(PatientID));
		list+=new DateStringAttribute(TagFromName(PatientBirthDate));
		list+=new CodeStringAttribute(TagFromName(PatientSex));
		list+=new ShortStringAttribute(TagFromName(StudyID));
		list+=new DateStringAttribute(TagFromName(StudyDate));
		list+=new TimeStringAttribute(TagFromName(StudyTime));
		list+=new PersonNameAttribute(TagFromName(ReferringPhysicianName),"^^^^");
		list+=new ShortStringAttribute(TagFromName(AccessionNumber));
		list+=new IntegerStringAttribute(TagFromName(SeriesNumber));
		list+=new IntegerStringAttribute(TagFromName(InstanceNumber));

		list+=new CodeStringAttribute(TagFromName(Modality),"OT");
		list+=new CodeStringAttribute(TagFromName(ConversionType),"WSD");

		list+=new LongStringAttribute(TagFromName(Manufacturer));
		list+=new CodeStringAttribute(TagFromName(PatientOrientation));
	}
	
	Raw_PixelDataSourceBase *pixeldatasrc=0;
	if (bits <= 8) {
		pixeldatasrc=new Raw_PixelDataSource_Byte(in,offset,rows,cols,frames,samples);
	}
	else if (littleendian) {
		pixeldatasrc=new Raw_PixelDataSource_LittleEndian(in,offset,rows,cols,frames,samples);
	}
	else if (bigendian) {
		pixeldatasrc=new Raw_PixelDataSource_BigEndian(in,offset,rows,cols,frames,samples);
	}
	Assert(pixeldatasrc);

	list+=new OtherUnspecifiedLargeAttribute(
		TagFromName(PixelData),
		pixeldatasrc,
		rows,
		cols,
		frames,
		samples,
		&transfersyntax,
		0,bitsAllocated,bitsStored,highBit);

	if (!usualManagedAttributeListWrite(list,dout,
		dicom_output_options,log,verbose)) return 1;

	return 0;
}

